core/arch/arm/plat-imx/psci.c

```diff
@@ -67,3 +71,67 @@ int psci_cpu_on(uint32_t core_idx, uint32_t entry,
 
 	return PSCI_RET_SUCCESS;
 }
+
+int psci_cpu_off(void)
+{
+	uint32_t core_id;
+	vaddr_t va = core_mmu_get_va(SRC_BASE, MEM_AREA_IO_SEC);
+
+	core_id = get_core_pos();
+
+	write32(~0, va + SRC_GPR1 + core_id * 8 + 4);
+
+	DMSG("core_id: %d\n", core_id);
+
+	psci_armv7_cpu_off();
+
+	imx_set_src_gpr(core_id, ~0);
+
+	thread_mask_exceptions(THREAD_EXCP_ALL);
+
+	asm volatile(
```

Comment:
This patch to enable usage of the suggested loop is in my opinion the lesser of two evils.

```
diff --git a/core/arch/arm/plat-imx/sub.mk b/core/arch/arm/plat-imx/sub.mk
index 41b4bbcd8b46..e42213caad9f 100644
--- a/core/arch/arm/plat-imx/sub.mk
+++ b/core/arch/arm/plat-imx/sub.mk
@@ -3,6 +3,7 @@ srcs-y += main.c imx-common.c
 
 srcs-$(CFG_PL310) += imx_pl310.c
 srcs-$(CFG_PSCI_ARM32) += psci.c
+cflags-psci.c-y += -Wno-suggest-attribute=noreturn
 
 ifneq (,$(filter y, $(CFG_MX6Q) $(CFG_MX6D) $(CFG_MX6DL)))
 srcs-y += a9_plat_init.S imx6.c
```